module/executiondatasync/pruner/pruner.go

```diff
@@ -0,0 +1,157 @@
+package pruner
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/module/component"
+	"github.com/onflow/flow-go/module/executiondatasync/tracker"
+	"github.com/onflow/flow-go/module/irrecoverable"
+	"github.com/onflow/flow-go/module/util"
+	"github.com/rs/zerolog"
+)
+
+const (
+	defaultHeightRangeTarget = uint64(400000)
+	defaultThreshold         = uint64(100000)
+)
+
+type Pruner struct {
+	storage tracker.Storage
+
+	fulfilledHeightsIn    chan<- interface{}
+	fulfilledHeightsOut   <-chan interface{}
+	thresholdChan         chan uint64
+	heightRangeTargetChan chan uint64
+
+	lastFulfilledHeight uint64
+	lastPrunedHeight    uint64
+	heightRangeTarget   uint64
+	threshold           uint64
+
+	logger  zerolog.Logger
+	metrics module.ExecutionDataPrunerMetrics
+
+	component.Component
+	cm *component.ComponentManager
+}
+
+type PrunerOption func(*Pruner)
+
+func WithHeightRangeTarget(heightRangeTarget uint64) PrunerOption {
+	return func(p *Pruner) {
+		p.heightRangeTarget = heightRangeTarget
+	}
+}
+
+func WithThreshold(threshold uint64) PrunerOption {
+	return func(p *Pruner) {
+		p.threshold = threshold
+	}
+}
+
+func NewPruner(logger zerolog.Logger, metrics module.ExecutionDataPrunerMetrics, storage tracker.Storage, opts ...PrunerOption) (*Pruner, error) {
+	lastPrunedHeight, err := storage.GetPrunedHeight()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get pruned height: %w", err)
+	}
+
+	fulfilledHeight, err := storage.GetFulfilledHeight()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get fulfilled height: %w", err)
+	}
+
+	fulfilledHeightsIn, fulfilledHeightsOut := util.UnboundedChannel()
+	fulfilledHeightsIn <- fulfilledHeight
+
+	p := &Pruner{
+		logger:                logger.With().Str("component", "execution_data_pruner").Logger(),
+		storage:               storage,
+		fulfilledHeightsIn:    fulfilledHeightsIn,
+		fulfilledHeightsOut:   fulfilledHeightsOut,
+		thresholdChan:         make(chan uint64),
+		heightRangeTargetChan: make(chan uint64),
+		lastFulfilledHeight:   fulfilledHeight,
+		lastPrunedHeight:      lastPrunedHeight,
+		heightRangeTarget:     defaultHeightRangeTarget,
+		threshold:             defaultThreshold,
+		metrics:               metrics,
+	}
+	p.cm = component.NewComponentManagerBuilder().
+		AddWorker(p.loop).
+		Build()
+	p.Component = p.cm
+
+	for _, opt := range opts {
+		opt(p)
+	}
+
+	return p, nil
+}
+
+func (p *Pruner) NotifyFulfilledHeight(height uint64) {
+	if util.CheckClosed(p.cm.ShutdownSignal()) {
```

Comment:
Because it's an unbounded channel, the in channel will always accept new items. So we should check the shutdownsignal first before sending to the channel, otherwise the send may succeed even if the shutdownsignal is closed